### PR TITLE
API allows for article published status

### PR DESCRIPTION
**Pull request by dorianbuck**
PT: https://www.pivotaltracker.com/story/show/179892938

## User Story
```
As a news API
In order to move articles through editorial pipeline
I would like to include a published category
```

## Tasks
- Create `published` column in `articles` table- Create user model
- Make `published` column contain boolean values, set default to false
- Modify `index` action in `articles_controller.rb` to only return published articles

## Screenshot

Skärmavbild 2021-10-18 kl  12 32 57
Skärmavbild 2021-10-18 kl  12 33 37

### CODE REVIEW

File: spec/models/article_spec.rb [hunk 1]
```diff
@@ -5,13 +5,15 @@
...
     it { is_expected.to have_db_column :category_name}
+    it { is_expected.to have_db_column :published}
   end
 
   describe 'Validations' do
     it { is_expected.to validate_presence_of :title }
     it { is_expected.to validate_presence_of :lede }
     it { is_expected.to validate_presence_of :body }
     it { is_expected.to validate_presence_of :category_name }
+    # it { is_expected.to validate_presence_of :published }
```
Comment (maxarvid) [inline spec/models/article_spec.rb]: Should this comment be here? / Should it be a comment?

Comment (dorianbuck) [reply spec/models/article_spec.rb]: This has been removed

File: spec/requests/api/article_published_status_spec.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      get "/api/articles/#{article.id}"
+    end
+
+    it { is_expected.to have_http_status 200 }
+
+    it 'is expected to return an article with the published status false' do
+      expect(response_json['article']['published']).to eq false
+    end
+  end
+end
```
Comment (offcast-dev) [inline spec/requests/api/article_published_status_spec.rb]: I don't see why you need separate test file for this. I think you can include asserting published status in already existing index or show spec

Comment (offcast-dev) [reply spec/requests/api/article_published_status_spec.rb]: Also I think if article is not published you probably don't want to see it returned by index action? I don't see not test nor functionality for that. If you planned that as separate feature I say you should include it here as just having published status is not a feature in of itself.

**Review by offcast-dev — COMMENTED** [review]
Refactor new specs to existing files. Add functionality so Article index action do not return articles that are not published

File: spec/requests/api/article_published_status_spec.rb (hunk 2 above)
Comment (dorianbuck) [reply spec/requests/api/article_published_status_spec.rb]: This has been adjusted. Thank you for your valuable input!
Skärmavbild 2021-10-18 kl  20 06 32

**Review by tochman — CHANGES_REQUESTED** [review]
The changes proposed are not connected to the title and user story. Please read the story carefully and consider what it is you are implementing. This is just an observation on my part. 

Besides that, please do not forget about thin controllers and fat models. Consider extracting the filtering of published articles to the model.

File: db/migrate/20211018095905_add_published_status_to_articles.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddPublishedStatusToArticles < ActiveRecord::Migration[6.1]
+  def change
+    add_column :articles, :published, :boolean
```
Comment (tochman) [inline db/migrate/20211018095905_add_published_status_to_articles.rb]: What is the default state of this attribute?

**Review by tochman — APPROVED** [review]

File: app/controllers/api/articles_controller.rb [hunk 4]
```diff
@@ -27,15 +26,16 @@ def show
...
 
-  def get_articles(category_name)
+  def get_published_articles(category_name)
+    articles_published = Article.where(published: true)
     if Category.pluck(:name).include? category_name
-      Article.where(category_name: params[:category_name])
+      articles_published.where(category_name: params[:category_name])
     else
-      Article.all
+      articles_published.all
```
Comment (offcast-dev) [inline app/controllers/api/articles_controller.rb]: Are you sure you need .all here?

Comment (offcast-dev) [reply app/controllers/api/articles_controller.rb]: I think it will work without it

**Review by offcast-dev — COMMENTED** [review]
.all may be redundant

File: app/controllers/api/articles_controller.rb (hunk 4 above)
Comment (dorianbuck) [reply app/controllers/api/articles_controller.rb]: It works without

File: .ruby-version [hunk 5]
```diff
@@ -1 +1 @@
-ruby-3.0.2
+ruby-2.5.1
```
Comment (lukeperrin10) [inline .ruby-version:1]: why the change?

File: spec/requests/api/articles_index_spec.rb [hunk 6]
```diff
@@ -57,4 +61,17 @@
...
+  describe 'when the article is not published' do
+    let!(:article_4) { create(:article, published: false)  }
+    before do
+      get "/api/articles/"
+    end
+
+    it { is_expected.to have_http_status 404 }
+
+    it 'is expected to NOT return an unpublished article' do
+      expect(response_json['message']).to eq "There are no articles in the database"
```
Comment (lukeperrin10) [inline spec/requests/api/articles_index_spec.rb]: This one doesn't make sense to me really. The assertion is that it is not expected to return an unpublished article. But the response message is "There are no articles in the database". My point is here you test for an article that has a status that is published: false. I would prefer to see a test where there are multiple articles with different published values that actually show it is only returning the published articles. This is only testing if the articles DB is empty.

**Review by lukeperrin10 — CHANGES_REQUESTED** [review]
Need to look at your tests. See comment

File: .ruby-version (hunk 5 above)
Comment (Fraseer) [reply .ruby-version:1]: Because Johnny cannot install 3.0.2 so we all revert so he could code along

Comment (lukeperrin10) [reply .ruby-version:1]: ok, my suggestion would be that this is set from the start of any project dependent on everyone's setup. Something to think about in future. Versions and dependencies changing in the middle of a project should be done cautiously.

**Review by lukeperrin10 — APPROVED** [review]
Ok, looks good now, just think about what you're testing.

**Pull request merged**